src/xdist/dsession.py

```diff
@@ -147,14 +153,49 @@ def loop_once(self) -> None:
         """Process one callback from one of the workers."""
         while 1:
             if not self._active_nodes:
-                # If everything has died stop looping
-                self.triggershutdown()
-                raise RuntimeError("Unexpectedly no active workers available")
+                # Worker teardown + recreation only occurs for CustomGroup Scheduler
+                if self.remake_nodes and isinstance(self.sched, CustomGroup):
+                    self.terminal.write_line("\n")
+                    self.remake_nodes = False
+
+                    num_workers = self.sched.dist_groups[self.sched.pending_groups[0]]['group_workers']
+                    if num_workers < len(self.nodemanager.specs):
+                        self.nodemanager.specs = self.nodemanager.specs[0:num_workers]
+                    self.trdist._status = {}
+
+                    new_nodes = self.nodemanager.setup_nodes(self.saved_put)
+                    self._active_nodes = set()
+                    self._active_nodes.update(new_nodes)
+                    self.sched.node2pending = {}
+                    self.sched.do_resched = True
+                else:
+                    # We aren't using CustomGroup scheduler and everything has died: stop looping
+                    self.triggershutdown()
+                    raise RuntimeError("Unexpectedly no active workers available")
             try:
                 eventcall = self.queue.get(timeout=2.0)
                 break
             except Empty:
+                # Custom logic for CustomGroup scheduler:
+                if isinstance(self.sched, CustomGroup):
+                    all_one = True
```

Comment:
I'd rename this variable if we want to share with the upstream - something like `all_nodes_finishing` or something.

Reply:
Renamed that, that makes more sense, thanks